### ISSUE TITLE
Can't make the config run on Kubernetes

### ISSUE THREAD

**Issue by jmereaux** [jitsi/docker-jitsi-meet#2]
Hi,

I apologize in advance because this is more a support request rather than an issue with the provided config.

I have tried to transform the docker-compose yaml to some kube deployment/services yaml files in order to deploy the config on IBM Cloud Kubernetes service.
It seems like the config is running, but when two peers connect to a room, the video feed only lasts for a few seconds and then drops.

Below is one suspicious entry I could grab from the videobridge logs:
```
JVB 2018-07-30 11:11:00.863 INFO: [1051] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 192.168.1.21:58164/udp/host (stream.RTP)
JVB 2018-07-30 11:11:00.886 INFO: [1051] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 192.168.1.21:58164/udp/host (stream.RTP)
JVB 2018-07-30 11:11:00.918 INFO: [1051] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:4443/tcp/host -> 192.168.1.21:9/tcp/host (stream.RTP)
JVB 2018-07-30 11:11:01.199 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 169.254.194.74:51441/udp/host (stream.RTP)
JVB 2018-07-30 11:11:01.221 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 192.168.23.1:51442/udp/host (stream.RTP)
JVB 2018-07-30 11:11:01.242 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 192.168.1.61:51443/udp/host (stream.RTP)
JVB 2018-07-30 11:11:01.263 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:10000/udp/host -> 169.254.194.74:51441/udp/host (stream.RTP)
JVB 2018-07-30 11:11:01.284 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:4443/tcp/host -> 169.254.194.74:9/tcp/host (stream.RTP)
JVB 2018-07-30 11:11:01.305 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:4443/tcp/host -> 192.168.23.1:9/tcp/host (stream.RTP)
JVB 2018-07-30 11:11:01.326 INFO: [1052] org.ice4j.ice.ConnectivityCheckClient.log() Pair failed: 172.30.204.17:4443/tcp/host -> 192.168.1.61:9/tcp/host (stream.RTP)
JVB 2018-07-30 11:17:45.612 INFO: [2753] org.jitsi.videobridge.Conference.log() CAT=stat expire_conf,conf_id=75082097e8838dbc duration=406,conf_count=0,ch_count=0,v_streams=0,conf_completed=1,no_payload_ch=4,no_transport_ch=4,total_ch=4,has_failed=true,has_partially_failed=true
```

To me, sounds like the problem is related to the videobridge, and especially with the UDP port 10000 that is not properly exposed outside of the cluster.

Could you please help me to diagnose what is wrong with my kube setup?

Thanks

**Comment by saghul** [comment]
> To me, sounds like the problem is related to the videobridge, and especially with the UDP port 10000 that is not properly exposed outside of the cluster.

I have no experience with k8s so I'm not sure how to help you. Not having port 10000 properly open is most likely the cause. You can test this by having 2 participants in the room (this should work thanks to p2p mode) and when a 3rd joins, audio and video go away. This usually means the video bridge is unreachable or that the media cannot be routed correctly.

**Comment by jmereaux** [comment]
Thanks for taking the time to answer @saghul 

I finally managed to expose the port UDP 10000 outside of the cluster on a fixed public IP address.
I used this IP address to fill up the configuration property `org.ice4j.ice.harvest.NAT_HARVESTER_PUBLIC_ADDRESS`

At first sight, seems like the setup now works. At least, the video feed between the participants (up to 4) does not end within 10 seconds.

However, I am a bit concerned about the TCP port 4443 on the videobridge. I managed to expose it outside of the cluster, but on another IP address (Kubernetes apparently refuses to expose mixed protocols UDP/TCP on the same LoadBalancer IP address)

What is this TCP port 4443 used for?
Do you think it is a problem if it is not exposed outside of the cluster on the same IP address as the UPD 10000?

Thanks

**Comment by damencho** [comment]
It is a fallback of using tcp for media in case of some restricted firewalls. And currently jvb will advertise it as the org.ice4j.ice.harvest.NAT_HARVESTER_PUBLIC_ADDRESS=<Public.IP.Address>, so using two addresses will be a problem and will not work.

**Comment by jmereaux** [comment]
@damencho you're saying that the TCP port is used as a fallback, so if I understand you correctly it can be not used at all, right?
What would be a symptom that could indicate something is wrong in my setup?

**Comment by damencho** [comment]
Yep, it is fallback. To test it you can join 2 participants in a meeting, then join a third one from a machine where its firwall restricts using udp (only leave port 53 udp for dns) and it should then fallback to using tcp.

**Comment by jmereaux** [comment]
ok this is clear, thanks.

One last question if I may: is this docker setup compatible with the Jitsi mobile app? Seems like I cannot connect to a conference hosted on my kubernetes cluster from the mobile app.

**Comment by damencho** [comment]
It is compatible, you just need a valid certificate for that deployment.

**Comment by jmereaux** [comment]
I have a valid SSL certificate. The ingress service in front of the kube cluster offers a domain name and the associate tls support.
When I connect to my jitsi meet url using my browser, the connection is indicated as secured (chrome trusts the certificate). But the mobile app does not work

**Comment by saghul** [comment]
What does the mobile app do? Red screen? What happens if you open your URL in the mobile browser?

**Comment by jmereaux** [comment]
If if open the url in the mobile browser, I get the page asking to either download the app or continue to the app.
I hit 'continue to the app', then the app opens and I get stuck waiting on a black screen (with my avatar as background) with some buttons at the back of the screen  (mute, hang up, ...)
Eventually, the red screen shows up stating that I have been disconnected, and it retries the connection

**Comment by saghul** [comment]
Hum. Can you get the logs with `adb logcat` assuming you're on Android?

**Comment by jmereaux** [comment]
I tested both on iOS and Android with the same results.
Here is the adb logcat I could capture.
[adb.log](https://github.com/jitsi/docker-jitsi-meet/files/2244705/adb.log)

**Comment by saghul** [comment]
Oh, I think I know what's happening. The BOSH URL in config.js is set to relative to allow for easy proxying, but mobile may not be liking that... can you edit your config.js and adjust it so that it's a full URL? (instead of just /http-bind)

**Comment by jmereaux** [comment]
Seems like it did the trick! Thanks

**Comment by saghul** [comment]
🤦‍♂️  I need to fix that. Thanks for the help troubleshooting this!

**Comment by jmereaux** [comment]
Thanks a lot to you for your support!

One last question and I will close the issue I promise.
Is there any configuration trick that would enable to use different public IP addresses according to the protocol (TCP or UDP) used to target the video bridge?

As I said above, Kubernetes refuses to expose 10000:UDP and 4443:TCP on the same public IP address.

**Comment by saghul** [comment]
I have no idea... At any rate, it's probably better to not expose 4443 and just use a TURN server instead.

**Comment by jmereaux** [comment]
Not sure to know what a TURN server is and how it could be added to my Kubernetes setup.

Anyway, my current setup seems to work properly now, putting aside that I will not work for people behind a firewall restricting outgoing UDP connections.

So I don't know if It's worth spending more time on exploring the TURN server path.

**Comment by jmereaux** [comment]
Just in case it may help someone else, here are the Kubernetes yaml files I used
[jitsi-kube-yaml.zip](https://github.com/jitsi/docker-jitsi-meet/files/2252544/jitsi-kube-yaml.zip)

**Comment by saghul** [comment]
@jmereaux thanks for sharing! Can you make it a PR? I'll try to take a look and adjust as necessary, I think this is something many people would like.

**Comment by jmereaux** [comment]
#3
